contracts/Stake/StakeManager.sol

```diff
@@ -152,286 +153,258 @@ contract StakeManager is Initializable, Ownable {
         require(
-            stash.undelegatesAt <= block.number,
-            "StakeManager:redelegateStash - Stash is not yet undelegated"
-        );
-        bytes32 lockId = keccak256(abi.encodePacked(REDELEGATION_LOCK_SELECTOR, stashes[_stashId].staker));
-        uint256 unlockBlock = locks[lockId].unlockBlock;
-        require(
-            unlockBlock <= block.number,
+            _unlockBlock <= block.number,
             "StakeManager:redelegateStash - Redelegation period is not yet complete"
         );
-        address updatedCluster = address(locks[lockId].iValue);
+        address _updatedCluster = address(locks[_lockId].iValue);
         require(
-            clusterRegistry.isClusterValid(updatedCluster),
+            clusterRegistry.isClusterValid(_updatedCluster),
             "StakeManager:redelegateStash - can't delegate to invalid cluster"
         );
-        bytes32[] memory tokens = stash.tokensDelegated;
-        uint256[] memory amounts = new uint256[](tokens.length);
-        for(uint256 i=0; i < tokens.length; i++) {
-            amounts[i] = stashes[_stashId].amount[tokens[i]].amount;
+        bytes32[] memory _tokens = rewardDelegators.getFullTokenList();
+        uint256[] memory _amounts = new uint256[](_tokens.length);
+        for(uint256 i=0; i < _tokens.length; i++) {
+            _amounts[i] = stashes[_stashId].amount[_tokens[i]];
         }
-        rewardDelegators.undelegate(msg.sender, stash.delegatedCluster, tokens, amounts);
-        rewardDelegators.delegate(msg.sender, updatedCluster, tokens, amounts);
-        stashes[_stashId].delegatedCluster = updatedCluster;
-        delete locks[lockId];
-        emit Redelegated(_stashId, updatedCluster);
+        rewardDelegators.undelegate(_stash.staker, _stash.delegatedCluster, _tokens, _amounts);
+        rewardDelegators.delegate(_stash.staker, _updatedCluster, _tokens, _amounts);
+        stashes[_stashId].delegatedCluster = _updatedCluster;
+        delete locks[_lockId];
+        emit Redelegated(_stashId, _updatedCluster);
     }
 
     function undelegateStash(bytes32 _stashId) public {
-        Stash memory stash = stashes[_stashId];
+        Stash memory _stash = stashes[_stashId];
         require(
-            stash.staker == msg.sender, 
+            _stash.staker == msg.sender,
             "StakeManager:undelegateStash - Only staker can undelegate stash"
         );
         require(
-            stash.delegatedCluster != address(0),
+            _stash.delegatedCluster != address(0),
             "StakeManager:undelegateStash - stash is not delegated to any cluster"
         );
-        require(
-            stash.undelegatesAt <= block.number,
-            "StakeManager:undelegateStash - stash is already waiting for undelegation"
-        );
-        uint256 waitTime = rewardDelegators.undelegationWaitTime();
+        uint256 _waitTime = rewardDelegators.undelegationWaitTime();
         // use + for gas savings as overflow can't happen
```

Comment:
May be not use + to ensure that even if  admin key  is compromised and waitTime can be arbitrarily changed. It is not possible for attacker to undelegate stash without waiting using overflow.